object-api-multipart.go

```diff
@@ -0,0 +1,495 @@
+	}
+	fileInfos, eof, e := o.storage.ListFiles(minioMetaVolume, path.Join(bucket, object, uploadID)+".", marker, false, maxParts)
+	if e != nil {
+		return result, probe.NewError(InvalidPart{})
+	}
+	for _, fileInfo := range fileInfos {
+		fileName := path.Base(fileInfo.Name)
+		splitResult := strings.Split(fileName, ".")
+		partNum, e := strconv.Atoi(splitResult[1])
+		if e != nil {
+			return result, probe.NewError(e)
+		}
+		md5sum := splitResult[2]
+		result.Parts = append(result.Parts, partInfo{
+			PartNumber:   partNum,
+			LastModified: fileInfo.ModTime,
+			ETag:         md5sum,
+			Size:         fileInfo.Size,
+		})
+		nextPartNumberMarker = partNum
+	}
+	result.Bucket = bucket
+	result.Object = object
+	result.UploadID = uploadID
+	result.PartNumberMarker = partNumberMarker
+	result.NextPartNumberMarker = nextPartNumberMarker
+	result.MaxParts = maxParts
+	result.IsTruncated = !eof
+	return result, nil
+}
+
+func (o objectAPI) CompleteMultipartUpload(bucket string, object string, uploadID string, parts []completePart) (ObjectInfo, *probe.Error) {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return ObjectInfo{}, probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if !IsValidObjectName(object) {
+		return ObjectInfo{}, probe.NewError(ObjectNameInvalid{Bucket: bucket, Object: object})
+	}
+	if status, e := o.isUploadIDExist(bucket, object, uploadID); e != nil {
+		return ObjectInfo{}, probe.NewError(e)
+	} else if !status {
+		return ObjectInfo{}, probe.NewError(InvalidUploadID{UploadID: uploadID})
+	}
+	fileWriter, e := o.storage.CreateFile(bucket, object)
+	if e != nil {
+		return ObjectInfo{}, nil
+	}
+	for _, part := range parts {
+		partSuffix := fmt.Sprintf("%s.%d.%s", uploadID, part.PartNumber, part.ETag)
+		var fileReader io.ReadCloser
+		fileReader, e = o.storage.ReadFile(minioMetaVolume, path.Join(bucket, object, partSuffix), 0)
+		if e != nil {
+			return ObjectInfo{}, probe.NewError(e)
+		}
+		_, e = io.Copy(fileWriter, fileReader)
+		if e != nil {
+			return ObjectInfo{}, probe.NewError(e)
+		}
+		e = fileReader.Close()
+		if e != nil {
+			return ObjectInfo{}, probe.NewError(e)
+		}
+	}
+	e = fileWriter.Close()
+	if e != nil {
+		return ObjectInfo{}, probe.NewError(e)
+	}
+	fi, e := o.storage.StatFile(bucket, object)
+	if e != nil {
+		return ObjectInfo{}, probe.NewError(e)
+	}
+	o.removeMultipartUpload(bucket, object, uploadID)
+	return ObjectInfo{
+		Bucket:  bucket,
+		Name:    object,
+		ModTime: fi.ModTime,
+		Size:    fi.Size,
+		IsDir:   false,
+	}, nil
+}
+
+func (o objectAPI) removeMultipartUpload(bucket, object, uploadID string) *probe.Error {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if !IsValidObjectName(object) {
+		return probe.NewError(ObjectNameInvalid{Bucket: bucket, Object: object})
+	}
+	marker := ""
+	for {
+		uploadIDFile := path.Join(bucket, object, uploadID)
```

Comment:
each upload id is identified by .minio/bucket/object/<uploadid>

object parts of the uploadid are: .minio/bucket/object/<uploadid>.partNum.md5sum